View open reviews on homepage

Merseleo opened the pull request:
Closes #121

## What I have made

- added list of open reviews on the _Homepage_ (making use of the _NamedList_ component)
  - a _PaperListEntry_ is shown, if the corresponding paper is either unreviewed or has the current status "Maybe" or has not enough reviews **and** belongs to a project, the user that is currently logged in, is member 
- added a skeleton component for a _PaperListEntry_

- added a new (project) controller method for requesting the current stage

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - [x] unit tests
  - [x] integration tests
  ~~- [ ] end-to-end tests~~ (not possible yet)
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: add skeleton component for the PaperListEntry component

Commit:
feat: show all open reviews from projects the current users participates on home page (using NamedList component)

Merseleo commented:
I just went through our meeting notes again and I read that the entries in this list should be grouped by project or that the corresponding project name should appear somewhere. What do you think about this at the moment? Should I still include this or should we put it at the end (and I'll open an issue for it) or should we ignore it completely?

Commit:
fix: add 'overflow-hidden' to homepage and concatenate open reviews from multiple projects correctly

Slartibartfass2 commented on tests/integration/paper-components/paper-list-entry-skeleton.test.ts:
```diff
@@ -0,0 +1,25 @@
+
+        // expect skeleton for paper id, paper title, authors
+        expect(screen.getAllByTestId("skeleton", { exact: false }).length).toBeGreaterThan(2);
+        unmount();
```
Is there a reason why you didn't just split this up into two tests?

Merseleo replied:
As they imo belongs together and the PaperListEntrySkeleton has a random component in it, which I want to test by explicitly enforcing one random case. 
But I changed the first test case to explicitly enforcing the second case and so I use two components to test both cases of the random decision and as I said, imo this should be one test instead of two. Do you think differently?

Slartibartfass2 replied:
I think you test two different things here: A skeleton with review and one without. IMO, that's two cases, but I'm also fine keeping it as is.

I didn't notice the randomness the first time, and we have to be careful with that so that there are no flaky tests. Good that you changed it.

Merseleo replied:
Ok and you are right, at the beginning the test was "flaky", because I ignored this randomness, but I think now it should work. And I can now understand, why it could make sense to split these tests.

Slartibartfass2 commented on src/routes/+page.ts:
```diff
@@ -58,5 +89,25 @@ export const load: PageLoad = () => {
+        });
+
+    // attach noop-catch to handle promise rejection correctly (see https://svelte.dev/docs/kit/load#Streaming-with-promises)
+    projectsMetadata.catch(() => {});
```
This is duplicated in line 58/59

Merseleo replied:
Thank you, deleted it.

Slartibartfass2 commented on src/routes/+page.ts:
```diff
@@ -27,25 +33,50 @@ async function requestProjectMetadata(project: Project): Promise<ProjectMetadata
+
+    for (let i = currentStage; i <= latestStage; i++) {
+        const allStageEntriesFromStageI: StageEntry[] = await projectController
+            .stage(currentStage)
```
That should be `i` instead of `currentStage`
```suggestion
            .stage(i)
```

Merseleo replied:
You are right, thank you!

Slartibartfass2 reviewed (changes requested):
Looks great. Just some smaller changes.

Commit:
test: correctly test both variants (outside test environment randomly chosen) of the PaperListEntrySkeleton

Commit:
fix: delete duplicate code and fix wrong index when requesting stage papers

Merseleo reviewed (commented):
Thank you for your feedback, hopefully everything is answered and done now.

Slartibartfass2 reviewed (dismissed):
Well done!

Just rebase and then we can merge it.

Commit:
refactor: split tests for PaperListEntrySkeleton and add papers with less than the required number of reviews to 'Open Reviews' list

Merseleo commented:
I rebased and fixed the "bug" that papers, which are not undecided, but not have the required number of reviews are not listed in the open reviews. Please have a look at this change.

If everything is ok, you can merge.

Slartibartfass2 commented on src/routes/+page.ts:
```diff
@@ -27,25 +33,59 @@ async function requestProjectMetadata(project: Project): Promise<ProjectMetadata
+
+    const numberOfRequiredReviews = await projectController
+        .get()
+        .then((project) => project.reviewDecisionMatrix.numberOfReviewers);
```
Here, we already have the whole project as an object, so we can just use the property directly
```suggestion
    const numberOfRequiredReviews = project.reviewDecisionMatrix.numberOfReviewers;
```

Merseleo replied:
Sorry, missed it. Thank you, I will change it.

Slartibartfass2 commented on src/routes/+page.ts:
```diff
@@ -27,25 +33,59 @@ async function requestProjectMetadata(project: Project): Promise<ProjectMetadata
+                }))
+                .filter(
+                    (paperInfo: PaperInfos) =>
+                        (isPaperUndecided(paperInfo.paper) ||
+                            paperInfo.paper.reviewData?.reviews.length) ??
+                        0 < numberOfRequiredReviews,
```
What are your thoughts on extracting this logic to a separate method, with an explanation of how you define an undecided paper?

Slartibartfass2 replied:
Maybe also add the mapping and the filtering function to the calls (`...getPapers().map(...)`)
I think this could help make this more descriptive.

Merseleo replied:
So you want a separate method for determining, whether a paper needs further reviews? I can add it.

Regarding your second comment: I do not really understand what you want to say?

Slartibartfass2 replied:
Initially, I thought of something like the following
```ts
const allStageEntriesFromStageI: StageEntry[] = await projectController
            .stage(i)
            .getPapers()
            .map((stageEntry) => ({
                paper: stageEntry.paper,
                projectId: project.id,
                /// TODO: request this information, e.g. from local store
                showReviewStatus: false,
            }))
            .filter((paperInfo) =>
                doesPaperNeedReview(paperInfo.paper, numberOfRequiredReviews),
            );
 ```
 but then I noticed that the first part is async and this doesn't work this way, so nevermind.

Merseleo replied:
Ok

Commit:
refactor: outsource check, whether a paper needs further review(s) and added test for the extracted code
